### DISCUSSION TITLE
Customized hotkey - long press CTRL V

### DISCUSSION THREAD

**Discussion by RecaiPoyraz** [hluk/CopyQ#1677] [Q&A]
Hi,
thank you for this great app.
I need help with a little function.
is that possible to add a customized shortcut like CTRL V longer press to show the clipboard manager windows?
for ex , when I press down the keys CTRL V in a half second than that will work as normal paste function but if I press longer like in 1-2 seconds than the clipboard manager windows will appear.
is this possible to implement in the program? maybe with a command?
thank you
br

**Comment by hluk** [comment]
It would not be perfect but you can override Ctrl+V global shortcut in the app and trigger `paste()` if keyboard modifiers are released before an interval.

<details>
<summary>Old version - see updated code in my next reply</summary>
```ini
[Command]
Command="
    copyq:
    var waitMs = 500;

    var start = Date.now();
    if (start - settings('override-paste-start') < waitMs + 250) {
        abort();
    }

    popup('Ctrl+V')

    try {
        settings('override-paste-start', start)

        while (
            queryKeyboardModifiers().length > 0
            && Date.now() - start < waitMs) {}

        if (queryKeyboardModifiers().length > 0) {
            popup('Ctrl+V overridden')
        } else {
            popup('Ctrl+V not overridden')
            paste();
        }
    } finally {
        settings('override-paste-start', null)
    }"
GlobalShortcut=ctrl+v
Icon=\xf11c
IsGlobalShortcut=true
Name=Override Paste
```
</details>

**Reply by RecaiPoyraz** [reply]
hi
thank you for your answer.
that didnt work and need more help now :)

after created a new global shortcut got below notifications from windows 10.
also copyq kept showing these notifications multiple times and copied the latest item several times.

![image](https://user-images.githubusercontent.com/6215778/121865077-d8bd2a00-ccfd-11eb-86d0-92436432e901.png)
![image](https://user-images.githubusercontent.com/6215778/121865090-db1f8400-ccfd-11eb-8792-135fd6552d52.png)

settings pages

![image](https://user-images.githubusercontent.com/6215778/121865194-fab6ac80-ccfd-11eb-9bff-7f890d0a4a9f.png)

**Reply by hluk** [reply]
I've updated the command - maybe this will be a bit better (those popups are just for debugging - if you don't want to see them, you can remove `popup(...)` lines from the code below):

```ini
[Command]
Command="
    copyq:
    var waitMs = 500;

    var start = Date.now();
    if (start - settings('override-paste-start') < 10000) {
        abort();
    }

    popup('Ctrl+V')

    try {
        settings('override-paste-start', start)
        
        while (
            queryKeyboardModifiers().length > 0
            && Date.now() - start < waitMs) {}
            
        if (queryKeyboardModifiers().length > 0) {
            popup('Ctrl+V overridden')
        } else {
            popup('Ctrl+V not overridden')
            paste();
        }
    } finally {
        while (queryKeyboardModifiers().length > 0) {}
        settings('override-paste-start', null)
    }"
GlobalShortcut=ctrl+v
Icon=\xf11c
IsGlobalShortcut=true
Name=Override Paste
```

**Reply by RecaiPoyraz** [reply]
hi 
thank you for the reply.
but the result is same.
that didnt work and got the below error msg.
br

![image](https://user-images.githubusercontent.com/6215778/122720760-242c8680-d270-11eb-80c1-26d4ea864fad.png)

**Reply by hluk** [reply]
Hmm, copying the command back to CopyQ works fine for me. You can check logs (F12 shortcut from main window) to find details of the error (especially line where it occurred).

**Reply by RecaiPoyraz** [reply]
hi
please see the log
thanks

`Note [2021-06-21 09:06:23.046] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:23.046] <Server-21356>:    
Note [2021-06-21 09:06:23.046] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:23.046] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:23.046] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:23.046] <Server-21356>:    eval
Note [2021-06-21 09:06:23.046] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:23.665] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:23.665] <Server-21356>:    
Note [2021-06-21 09:06:23.665] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:23.665] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:23.665] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:23.665] <Server-21356>:    eval
Note [2021-06-21 09:06:23.665] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.567] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.567] <Server-21356>:    
Note [2021-06-21 09:06:24.567] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.567] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.567] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.567] <Server-21356>:    eval
Note [2021-06-21 09:06:24.567] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.568] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.568] <Server-21356>:    
Note [2021-06-21 09:06:24.568] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.568] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.568] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.568] <Server-21356>:    eval
Note [2021-06-21 09:06:24.568] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.569] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.569] <Server-21356>:    
Note [2021-06-21 09:06:24.569] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.569] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.569] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.569] <Server-21356>:    eval
Note [2021-06-21 09:06:24.569] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.570] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.570] <Server-21356>:    
Note [2021-06-21 09:06:24.570] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.570] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.570] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.570] <Server-21356>:    eval
Note [2021-06-21 09:06:24.570] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.571] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.571] <Server-21356>:    
Note [2021-06-21 09:06:24.571] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.571] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.571] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.571] <Server-21356>:    eval
Note [2021-06-21 09:06:24.571] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.572] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.572] <Server-21356>:    
Note [2021-06-21 09:06:24.572] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.572] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.572] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.572] <Server-21356>:    eval
Note [2021-06-21 09:06:24.572] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.573] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.573] <Server-21356>:    
Note [2021-06-21 09:06:24.573] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.573] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.573] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.573] <Server-21356>:    eval
Note [2021-06-21 09:06:24.573] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.574] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.574] <Server-21356>:    
Note [2021-06-21 09:06:24.574] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.574] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.574] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.574] <Server-21356>:    eval
Note [2021-06-21 09:06:24.574] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.576] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.576] <Server-21356>:    
Note [2021-06-21 09:06:24.576] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.576] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.576] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.576] <Server-21356>:    eval
Note [2021-06-21 09:06:24.576] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.582] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.582] <Server-21356>:    
Note [2021-06-21 09:06:24.582] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.582] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.582] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.582] <Server-21356>:    eval
Note [2021-06-21 09:06:24.582] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.590] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.590] <Server-21356>:    
Note [2021-06-21 09:06:24.590] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.590] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.590] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.590] <Server-21356>:    eval
Note [2021-06-21 09:06:24.590] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.725] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.725] <Server-21356>:    
Note [2021-06-21 09:06:24.725] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.725] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.725] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.725] <Server-21356>:    eval
Note [2021-06-21 09:06:24.725] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.760] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.760] <Server-21356>:    
Note [2021-06-21 09:06:24.760] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.760] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.760] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.760] <Server-21356>:    eval
Note [2021-06-21 09:06:24.760] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.781] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.781] <Server-21356>:    
Note [2021-06-21 09:06:24.781] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.781] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.781] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.781] <Server-21356>:    eval
Note [2021-06-21 09:06:24.781] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.828] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.828] <Server-21356>:    
Note [2021-06-21 09:06:24.828] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.828] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.828] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.828] <Server-21356>:    eval
Note [2021-06-21 09:06:24.828] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.892] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.892] <Server-21356>:    
Note [2021-06-21 09:06:24.892] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.892] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.892] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.892] <Server-21356>:    eval
Note [2021-06-21 09:06:24.892] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.898] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.898] <Server-21356>:    
Note [2021-06-21 09:06:24.898] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.898] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.898] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.898] <Server-21356>:    eval
Note [2021-06-21 09:06:24.898] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.923] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.923] <Server-21356>:    
Note [2021-06-21 09:06:24.923] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.923] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.923] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.923] <Server-21356>:    eval
Note [2021-06-21 09:06:24.923] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.930] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.930] <Server-21356>:    
Note [2021-06-21 09:06:24.930] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.930] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.930] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.930] <Server-21356>:    eval
Note [2021-06-21 09:06:24.930] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.942] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.942] <Server-21356>:    
Note [2021-06-21 09:06:24.942] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.942] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.942] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.942] <Server-21356>:    eval
Note [2021-06-21 09:06:24.942] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.949] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.949] <Server-21356>:    
Note [2021-06-21 09:06:24.949] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.949] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.949] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.949] <Server-21356>:    eval
Note [2021-06-21 09:06:24.949] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.971] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.971] <Server-21356>:    
Note [2021-06-21 09:06:24.971] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.971] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.971] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.971] <Server-21356>:    eval
Note [2021-06-21 09:06:24.971] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.978] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.978] <Server-21356>:    
Note [2021-06-21 09:06:24.978] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.978] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.978] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.978] <Server-21356>:    eval
Note [2021-06-21 09:06:24.978] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.983] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.983] <Server-21356>:    
Note [2021-06-21 09:06:24.983] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.983] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.983] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.983] <Server-21356>:    eval
Note [2021-06-21 09:06:24.983] <Server-21356>:    --- end backtrace ---
Note [2021-06-21 09:06:24.986] <Server-21356>: Exception in command "Override Paste": ScriptError: SyntaxError: Unexpected token `}'
Note [2021-06-21 09:06:24.986] <Server-21356>:    
Note [2021-06-21 09:06:24.986] <Server-21356>:    --- backtrace ---
Note [2021-06-21 09:06:24.986] <Server-21356>:    @eval code:16
Note [2021-06-21 09:06:24.986] <Server-21356>:    eval:var waitMs = 500; var start...
Note [2021-06-21 09:06:24.986] <Server-21356>:    eval
Note [2021-06-21 09:06:24.986] <Server-21356>:    --- end backtrace ---`

**Reply by RecaiPoyraz** [reply]
here you have a gif for how I tried
![copyq](https://user-images.githubusercontent.com/6215778/122725967-d450be00-d275-11eb-96b3-31bf020a2a83.gif)

**Reply by hluk** [reply]
[Here is how to add the command to CopyQ](http://copyq.readthedocs.io/en/latest/faq.html#how-to-load-shared-commands-and-share-them).

I've removed the popups and fixed the command according to your question.

```ini
[Command]
Command="
    copyq:
    var waitMs = 500;

    var start = Date.now();
    if (start - settings('override-paste-start') < 10000) {
        abort();
    }

    try {
        settings('override-paste-start', start)
        
        while (
            queryKeyboardModifiers().length > 0
            && Date.now() - start < waitMs) {}
            
        if (queryKeyboardModifiers().length > 0) {
            show();
        } else {
            paste();
        }
    } finally {
        while (queryKeyboardModifiers().length > 0) {}
        settings('override-paste-start', null)
    }"
GlobalShortcut=ctrl+v
Icon=\xf11c
IsGlobalShortcut=true
Name=Override Paste
```

**Comment by RecaiPoyraz** [comment]
hi
the below code you sent worked well.
thank you for your time and effort :)

`[Command]
Command="
    copyq:
    var waitMs = 500;

    var start = Date.now();
    if (start - settings('override-paste-start') < 10000) {
        abort();
    }

    try {
        settings('override-paste-start', start)
        
        while (
            queryKeyboardModifiers().length > 0
            && Date.now() - start < waitMs) {}
            
        if (queryKeyboardModifiers().length > 0) {
            show();
        } else {
            paste();
        }
    } finally {
        while (queryKeyboardModifiers().length > 0) {}
        settings('override-paste-start', null)
    }"
GlobalShortcut=ctrl+v
Icon=\xf11c
IsGlobalShortcut=true
Name=Override Paste`
